pkg/kubevirt/data_volume.go

```diff
@@ -0,0 +1,149 @@
+	cdi "kubevirt.io/containerized-data-importer/pkg/apis/core/v1alpha1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// ClientFactory creates a client from the kubeconfig saved in the "kubeconfig" field of the given secret.
+type ClientFactory interface {
+	// GetClient creates a client from the kubeconfig saved in the "kubeconfig" field of the given secret.
+	// It also returns the namespace of the kubeconfig's current context.
+	GetClient(kubeconfig []byte) (client.Client, string, error)
+}
+
+// ClientFactoryFunc is a function that implements ClientFactory.
+type ClientFactoryFunc func(kubeconfig []byte) (client.Client, string, error)
+
+// GetClient creates a client from the kubeconfig saved in the "kubeconfig" field of the given secret.
+// It also returns the namespace of the kubeconfig's current context.
+func (f ClientFactoryFunc) GetClient(kubeconfig []byte) (client.Client, string, error) {
+	return f(kubeconfig)
+}
+
+// DataVolumeManager manages Kubevirt DataVolume operations.
+type DataVolumeManager interface {
+	// CreateVolume creates a new kubevirt Data Volume from the data volume specs and in the passed namespace.
+	CreateVolume(ctx context.Context, kubeconfig []byte, dataVolume *cdi.DataVolume) error
+	// GetVolume fetches the specified volume by the passed name and namespace and return DataVolumeNotFoundError error in case of
+	// not found object error.
+	GetVolume(ctx context.Context, kubeconfig []byte, name string) (*cdi.DataVolume, error)
+	// ListDataVolumes lists all the Data Volumes which exists in the passed namespace.
+	ListDataVolumes(ctx context.Context, kubeconfig []byte, listOpts ...client.ListOption) (*cdi.DataVolumeList, error)
+	// DeleteVolume delete the DataVolume based on the passed name and namespace.
+	DeleteVolume(ctx context.Context, kubeconfig []byte, name string) error
+}
+
+type defaultDataVolumeManager struct {
+	client ClientFactory
+	logger logr.Logger
+}
+
+// NewDefaultDataVolumeManager creates a new default manager with a k8s controller-runtime based on the sent kubecomfig.
+func NewDefaultDataVolumeManager(client ClientFactory) (DataVolumeManager, error) {
+	return &defaultDataVolumeManager{
+		client: client,
+		logger: log.Log.WithName("kubevirt-data-volume-manager"),
+	}, nil
+}
+
+// CreateVolume creates a new kubevirt Data Volume from the data volume specs and in the passed namespace.
+func (d *defaultDataVolumeManager) CreateVolume(ctx context.Context, kubeconfig []byte, dataVolume *cdi.DataVolume) error {
+	c, namespace, err := d.client.GetClient(kubeconfig)
+	if err != nil {
+		return fmt.Errorf("failed to create kubevirt client: %v", err)
+	}
+
+	_, err = d.GetVolume(ctx, kubeconfig, dataVolume.Name)
+	if err != nil {
+		if IsDataVolumeNotFoundError(err) {
+			dataVolume.Namespace = namespace
+			return c.Create(ctx, dataVolume)
+		}
+
+		return fmt.Errorf("failed to create data volume: %v", err)
+	}
```

Comment:
As previously suggested:

* Rename this method to `CreateOrUpdateVolume`.
* Use `controllerutil.CreateOrUpdate` inside this method to create or update the volume.

You don't need to get the volume and create if it doesn't exist, the logic in `controllerutil.CreateOrUpdate` will properly do it for you, see how it's done in `pkg/controllers/infrastructure/actuator_reconcile.go`